src/include/migraphx/op/nonmaxsuppression.hpp

```diff
@@ -220,30 +226,42 @@ struct nonmaxsuppression
 
     // filter boxes below score_threshold
     template <class T>
-    std::priority_queue<std::pair<double, int64_t>>
+    std::vector<std::pair<double, int64_t>>
     filter_boxes_by_score(T scores_start, std::size_t num_boxes, double score_threshold) const
     {
-        std::priority_queue<std::pair<double, int64_t>> boxes_heap;
-        auto insert_to_boxes_heap =
-            make_function_output_iterator([&](const auto& x) { boxes_heap.push(x); });
+        std::vector<std::pair<double, int64_t>> boxes_heap;
         int64_t box_idx = 0;
-        transform_if(
-            scores_start,
-            scores_start + num_boxes,
-            insert_to_boxes_heap,
-            [&](auto sc) {
-                box_idx++;
-                return sc >= score_threshold;
-            },
-            [&](auto sc) { return std::make_pair(sc, box_idx - 1); });
+
+        if(score_threshold > 0.0)
+        {
+            transform_if(
+                scores_start,
+                scores_start + num_boxes,
+                std::back_inserter(boxes_heap),
+                [&](auto sc) {
+                    box_idx++;
+                    return sc >= score_threshold;
+                },
+                [&](auto sc) { return std::make_pair(sc, box_idx - 1); });
+        }
+        else
+        { // score is irrelevant, just push into boxes_heap and make a score-index pair
+            std::transform(scores_start,
+                           scores_start + num_boxes,
+                           std::back_inserter(boxes_heap),
+                           [&](auto sc) {
+                               box_idx++;
+                               return std::make_pair(sc, box_idx - 1);
+                           });
+        }
+        par_sort(boxes_heap.begin(), boxes_heap.end(), std::greater<std::pair<double, int64_t>>{});
```

Comment:
A dumb question. The onnx spec doesn't specify sorting is required.. I must be missing something. Thanks.

Reply:
The sorting is done to determine the intersection of union of the boxes between selected boxes. This sort allows us to find the boxes with the greatest overlap in the set of all detected boxes

Reply:
Thanks, but that is not a reference document, Ted. I am merely aiming that we should do the bare-minimum, as per the spec, and not do costly overheads, which are not specified. I understand, this has nothing to do with just your set of changes, and sort of changes even the prior logic..

Reply:
I think we should let sort stay. I have looked at TF and at https://pytorch.org/vision/stable/generated/torchvision.ops.nms.html, and they specify sorted order. _"int64 tensor with the indices of the elements that have been kept by NMS, sorted in decreasing order of scores"_

The onnx spec is not very clear about it.

Reply:
TF: _Greedily selects a subset of bounding boxes in descending order of score_. https://www.tensorflow.org/api_docs/python/tf/image/non_max_suppression